File: day08/b.ts
const regex = /(?<Key>\w+) = \((?<L>\w+),\s(?<R>\w+)\)/;

const file = await Deno.readTextFile("./input.txt");

const [path, map_text] = file.split("\n\n");

const mapIV: {
  [key: string]: string[];
} = {};

const lines = map_text
  .split("\n")
  .map((l) => {
    const groups = l.match(regex)?.groups;
    return {
      key: groups?.Key ?? "oops",
      value: [groups?.L ?? "oops", groups?.R ?? "oops"],
    };
  })
  .reduce((acc, cur) => {
    acc[cur.key] = cur.value;
    return acc;
  }, mapIV);

function get(key: string, value: string) {
  return lines[key][+!(value == "L")];
}
const pathArr = path.split("");

const starts = Object.keys(lines).filter((l) => l.endsWith("A"));
function solve(s: string) {
  let i = 0;
  let current = s;

  while (!current.endsWith("Z")) {
    current = get(current, pathArr[i % pathArr.length]);
    i++;
  }

  return i;
}
const lcm = (a: number, b: number) => (a * b) / gcd(a, b);
const gcd = (a: number, b: number): number => (b ? gcd(b, a % b) : a);

const res = (
  await Promise.all(
    starts.map(
      (s) =>
        new Promise((resolve, _) => {
          resolve(solve(s));
        }),
    ),
  )
).map((n) => n as number).reduce((acc, cur) => lcm(acc, cur), 1);

console.log(res);

File: day08/b.go
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"
	"sync"
)

func Part2() {
	file, err := os.ReadFile("./input.txt")
	if err != nil {
		panic(err)
	}

	re := regexp.MustCompile(`(?P<K>\w+) = \((?P<L>\w+), (?P<R>\w+)\)`)
	data := strings.Split(strings.TrimSpace(string(file)), "\n\n")
	path := strings.Split(data[0], "")
	rawMap := data[1]
	parsedMap := map[string]MapEntry{}

	lines := strings.Split(rawMap, "\n")
	keys := make([]string, len(lines))
	for i, v := range lines {
		submatches := re.FindStringSubmatch(v)
		names := re.SubexpNames()
		result := map[string]string{}

		for i, name := range names {
			if i != 0 && name != "" {
				result[name] = submatches[i]
			}
		}
		parsedMap[result["K"]] = MapEntry{
			L: result["L"],
			R: result["R"],
		}
		keys[i] = result["K"]
	}

	solve := func(start string, data map[string]MapEntry, path []string) int {
		cur := start
		counter := 0
		for !strings.HasSuffix(cur, "Z") {
			nextDir := path[counter%len(path)]
			next := data[cur]
			if nextDir == "L" {
				cur = next.L
			} else {
				cur = next.R
			}
			counter++
		}
		return counter
	}

	starts := make([]string, 0, len(keys))

	for _, v := range keys {
		if strings.HasSuffix(v, "A") {
			starts = append(starts, v)
		}
	}

	var wg sync.WaitGroup

	wg.Add(len(starts))

	results := make(chan int, len(starts))

	for _, start := range starts {
		go func(s string, c chan int, data map[string]MapEntry, path []string) {
			defer wg.Done()
			res := solve(s, data, path)
			c <- res

		}(start, results, parsedMap, path)
	}

	go func() {
		wg.Wait()
		close(results)
	}()

	total := 1
	for res := range results {
		total = lcm(total, res)
	}

	fmt.Printf("Part2 {%d}", total)
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func lcm(a, b int) int {
	return (a * b) / gcd(a, b)
}
